<a href="https://colab.research.google.com/github/LastCodeBender42/TensorFlow-Certification-Prep/blob/main/01_Basic_Tensor_Operations_Ex.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import tensorflow as tf
import numpy as np

# 1. **Basic Tensor Operations**

### **Exercise 1.1**: Create a TensorFlow tensor with a shape of `(3, 3)` and initialize it with random values. Perform basic operations such as addition, subtraction, multiplication, and division on this tensor.

```python
import tensorflow as tf

tensor = tf.random.uniform((3, 3))

# Basic operations
add_result = tensor + 1
sub_result = tensor - 1
mul_result = tensor * 2
div_result = tensor / 2
```

In [ ]:
tensor = tf.random.uniform((3, 3))
print(tensor)

tf.Tensor(
[[0.5447719  0.34592664 0.9700761 ]
 [0.02648783 0.6058005  0.5102856 ]
 [0.24757922 0.24412417 0.8933885 ]], shape=(3, 3), dtype=float32)


In [ ]:
addition = tensor + 1
print(addition)

tf.Tensor(
[[1.5447719 1.3459266 1.9700761]
 [1.0264878 1.6058005 1.5102856]
 [1.2475792 1.2441242 1.8933885]], shape=(3, 3), dtype=float32)


In [ ]:
subtraction = tensor - 1
print(subtraction)

tf.Tensor(
[[-0.4552281  -0.65407336 -0.02992392]
 [-0.9735122  -0.3941995  -0.48971438]
 [-0.7524208  -0.7558758  -0.10661149]], shape=(3, 3), dtype=float32)


In [ ]:
multiplication = tensor * 2
print(multiplication)

tf.Tensor(
[[1.0895438  0.6918533  1.9401522 ]
 [0.05297565 1.211601   1.0205712 ]
 [0.49515843 0.48824835 1.786777  ]], shape=(3, 3), dtype=float32)


In [ ]:
division = tensor / 2
print(division)

tf.Tensor(
[[0.27238595 0.17296332 0.48503804]
 [0.01324391 0.30290025 0.2551428 ]
 [0.12378961 0.12206209 0.44669425]], shape=(3, 3), dtype=float32)


### **Exercise 1.2**: Create a tensor with a shape of `(4, 4)` and use TensorFlow functions to reshape it into a `(2, 8)` tensor. Verify the reshaped tensor’s shape.

```python
import tensorflow as tf

tensor = tf.random.uniform((4, 4))

# Reshape
reshaped_tensor = tf.reshape(tensor, (2, 8))
```


In [ ]:
tensor = tf.random.uniform((4, 4))
print(tensor)

tf.Tensor(
[[0.34866953 0.77890337 0.571784   0.8423252 ]
 [0.479416   0.09513104 0.17833817 0.96039426]
 [0.0445652  0.39942813 0.65387285 0.28827584]
 [0.7688081  0.71405697 0.8793558  0.8422322 ]], shape=(4, 4), dtype=float32)


In [ ]:
reshaped_tensor = tf.reshape(tensor, (2, 8))
print(reshaped_tensor)

tf.Tensor(
[[0.34866953 0.77890337 0.571784   0.8423252  0.479416   0.09513104
  0.17833817 0.96039426]
 [0.0445652  0.39942813 0.65387285 0.28827584 0.7688081  0.71405697
  0.8793558  0.8422322 ]], shape=(2, 8), dtype=float32)


### **Exercise 1.3**: Implement broadcasting by adding a 1D tensor of shape `(3,)` to a 2D tensor of shape `(3, 3)`.

```python
import tensorflow as tf

tensor_2d = tf.random.uniform((3, 3))
tensor_1d = tf.constant([1, 2, 3])

# Broadcasting
result = tensor_2d + tensor_1d
```

In [ ]:
tensor_2d = tf.random.uniform((3, 3),dtype=tf.float32)
tensor_1d = tf.constant([1, 2, 3],dtype=tf.float32)

# Broadcasting
result = tensor_2d + tensor_1d

print("\nThe 2d tensor:")
print(tensor_2d)
print("\nThe 1d tensor:")
print(tensor_1d)
print("\nThe result of the broadcasting:")
print(result)


The 2d tensor:
tf.Tensor(
[[0.68652344 0.07966292 0.7592844 ]
 [0.6899688  0.7968639  0.4561032 ]
 [0.9190253  0.17476869 0.28766477]], shape=(3, 3), dtype=float32)

The 1d tensor:
tf.Tensor([1. 2. 3.], shape=(3,), dtype=float32)

The result of the broadcasting:
tf.Tensor(
[[1.6865234 2.0796628 3.7592845]
 [1.6899688 2.796864  3.4561033]
 [1.9190253 2.1747687 3.287665 ]], shape=(3, 3), dtype=float32)


### **Exercise 1.4**: Create a 3D tensor of shape `(2, 3, 4)` with random values and compute its mean along the first axis.

```python
import tensorflow as tf

tensor = tf.random.uniform((2, 3, 4))
mean_result = tf.reduce_mean(tensor, axis=0)
print(mean_result)
```

In [ ]:
tensor = tf.random.uniform((2,3,4),dtype=tf.float32)
print(tensor)
print("\n")
mean_result = tf.reduce_mean(tensor, axis=0)
print(mean_result)

tf.Tensor(
[[[0.5542735  0.314996   0.638111   0.14362872]
  [0.29933417 0.11891592 0.34602952 0.29040265]
  [0.22507727 0.93163526 0.32360995 0.76137304]]

 [[0.98418534 0.03321695 0.04477477 0.81922424]
  [0.86157894 0.02341616 0.95956933 0.40995634]
  [0.56845367 0.16571736 0.78648114 0.75783205]]], shape=(2, 3, 4), dtype=float32)


tf.Tensor(
[[0.7692294  0.17410648 0.34144288 0.48142648]
 [0.58045655 0.07116604 0.6527994  0.3501795 ]
 [0.39676547 0.5486763  0.55504555 0.75960255]], shape=(3, 4), dtype=float32)


### **Exercise 1.5**: Create two tensors of shape `(5, 5)` and perform element-wise subtraction. Ensure that both tensors are of the same data type.

```python
import tensorflow as tf

tensor1 = tf.random.uniform((5, 5))
tensor2 = tf.random.uniform((5, 5))
sub_result = tensor1 - tensor2
print(sub_result)
```

In [ ]:
tensor_1 = tf.random.uniform((5,5),dtype=tf.float32)
tensor_2 = tf.random.uniform((5,5),dtype=tf.float32)
tensor = tensor_1 - tensor_2
print(tensor)

tf.Tensor(
[[ 0.03186798 -0.83264554  0.09980607  0.4052782  -0.20020962]
 [ 0.5847888  -0.20748603 -0.07569504 -0.3127364   0.3367715 ]
 [-0.05200255  0.75205183 -0.28210914  0.59101033 -0.09743357]
 [-0.4948579  -0.38992333 -0.57157314  0.46262193  0.03320003]
 [-0.05695426 -0.04780149  0.04279864 -0.44929898  0.03484702]], shape=(5, 5), dtype=float32)


### **Exercise 1.6**: Create a tensor of shape `(3, 3)` initialized with zeros and another tensor of the same shape initialized with ones. Perform element-wise multiplication.

```python
import tensorflow as tf

zeros_tensor = tf.zeros((3, 3))
ones_tensor = tf.ones((3, 3))
mul_result = zeros_tensor * ones_tensor
print(mul_result)
```

In [ ]:
zeros_tensor = tf.zeros((3, 3))
ones_tensor = tf.ones((3, 3))
mul_result = zeros_tensor * ones_tensor
print(mul_result)

tf.Tensor(
[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]], shape=(3, 3), dtype=float32)


#### **NOTE:** Alternatively I could have used:


In [ ]:
shape = (3, 3)
zeros_tensor = tf.fill(shape, 0)
shape = (3, 3)
ones_tensor = tf.fill(shape, 1)
result = zeros_tensor * ones_tensor
print(result)

tf.Tensor(
[[0 0 0]
 [0 0 0]
 [0 0 0]], shape=(3, 3), dtype=int32)


### **Exercise 1.7**: Create a tensor of shape `(4, 4)` with random integers between 1 and 10. Find the maximum value in each row.

```python
import tensorflow as tf

tensor = tf.random.uniform((4, 4), minval=1, maxval=10, dtype=tf.int32)
max_per_row = tf.reduce_max(tensor, axis=1)
print(max_per_row)
```

In [ ]:
tensor = tf.random.uniform((4, 4), minval=1, maxval=10, dtype=tf.int32)
print(tensor)
max_values = tf.reduce_max(tensor, axis=1)
print("\n")
print(max_values)

tf.Tensor(
[[2 1 5 7]
 [2 5 5 2]
 [5 8 8 7]
 [7 9 9 4]], shape=(4, 4), dtype=int32)


tf.Tensor([7 5 8 9], shape=(4,), dtype=int32)


### **Exercise 1.8**: Create a tensor of shape `(3, 3)` and transpose it. Verify the shape of the transposed tensor.

```python
import tensorflow as tf

tensor = tf.random.uniform((3, 3))
transposed_tensor = tf.transpose(tensor)
print(transposed_tensor)
```

In [ ]:
|tensor = tf.random.uniform((3, 3))
print(tensor)
print("\n")
transposed_tensor = tf.transpose(tensor)
print(transposed_tensor)

tf.Tensor(
[[0.40123785 0.40710688 0.04617584]
 [0.49625254 0.08450532 0.5161176 ]
 [0.821808   0.61926293 0.07169294]], shape=(3, 3), dtype=float32)


tf.Tensor(
[[0.40123785 0.49625254 0.821808  ]
 [0.40710688 0.08450532 0.61926293]
 [0.04617584 0.5161176  0.07169294]], shape=(3, 3), dtype=float32)


### **Exercise 1.9**: Create a tensor of shape `(2, 2, 2)` and compute the sum of elements across the last axis.

```python
import tensorflow as tf

tensor = tf.random.uniform((2, 2, 2))
sum_result = tf.reduce_sum(tensor, axis=-1)
print(sum_result)
```

In [ ]:
tensor = tf.random.uniform((2,2,2))
print(tensor)
print("\n")
sum_result = tf.reduce_sum(tensor, axis=-1)
print(sum_result)

tf.Tensor(
[[[0.7357453  0.5311099 ]
  [0.5906867  0.5969068 ]]

 [[0.5098833  0.09803319]
  [0.5097177  0.47283077]]], shape=(2, 2, 2), dtype=float32)


tf.Tensor(
[[1.2668552 1.1875935]
 [0.6079165 0.9825485]], shape=(2, 2), dtype=float32)


### **Exercise 1.10**: Create a tensor with shape `(4, 4)` and use slicing to extract a sub-tensor of shape `(2, 2)` from the top-left corner.

```python
import tensorflow as tf

tensor = tf.random.uniform((4, 4))
sub_tensor = tensor[:2, :2]
print(sub_tensor)
```

In [ ]:
tensor = tf.random.uniform((4, 4))
print(tensor)
print("\n")
sub_tensor = tensor[:2, :2]
print(sub_tensor)

tf.Tensor(
[[0.9508574  0.03097332 0.80580556 0.14855552]
 [0.49364114 0.02638853 0.42838562 0.7785337 ]
 [0.23110032 0.7297523  0.11707997 0.4380455 ]
 [0.06297266 0.15089965 0.6792463  0.3209175 ]], shape=(4, 4), dtype=float32)


tf.Tensor(
[[0.9508574  0.03097332]
 [0.49364114 0.02638853]], shape=(2, 2), dtype=float32)


In [ ]:
sliced_tensor = tensor[1:3,1:3]
print(sliced_tensor)

tf.Tensor(
[[0.02638853 0.42838562]
 [0.7297523  0.11707997]], shape=(2, 2), dtype=float32)


In [ ]:
sliced_tensor = tensor[:4,2:4]
print(sliced_tensor)

tf.Tensor(
[[0.80580556 0.14855552]
 [0.42838562 0.7785337 ]
 [0.11707997 0.4380455 ]
 [0.6792463  0.3209175 ]], shape=(4, 2), dtype=float32)


In [ ]:
indices = tf.range(4)  # [0, 1, 2, 3]
new_tensor = tf.gather_nd(tensor, tf.stack((indices, indices), axis=1))
print(new_tensor)
print("\n")
# Reshaping to 1x4 tensor
new_tensor = tf.reshape(new_tensor, (1, 4))
print(new_tensor)

tf.Tensor([0.9508574  0.02638853 0.11707997 0.3209175 ], shape=(4,), dtype=float32)


tf.Tensor([[0.9508574  0.02638853 0.11707997 0.3209175 ]], shape=(1, 4), dtype=float32)


### **Exercise 1.11**: Create a tensor of shape `(2, 3)` and use `tf.concat` to concatenate it with another tensor of shape `(2, 2)` along the second axis. Ensure the resulting tensor's shape is correct.

```python
import tensorflow as tf

tensor1 = tf.random.uniform((2, 3))
tensor2 = tf.random.uniform((2, 2))
concat_result = tf.concat([tensor1, tensor2], axis=1)
print(concat_result)
```


In [ ]:
tensor1 = tf.random.uniform((2, 3))
print(tensor1)
print("\n")
tensor2 = tf.random.uniform((2, 2))
print(tensor2)
print("\n")
concat_result = tf.concat([tensor1, tensor2], axis=1)
print(concat_result)

tf.Tensor(
[[0.13274968 0.00210023 0.34535766]
 [0.887748   0.14301729 0.56532407]], shape=(2, 3), dtype=float32)


tf.Tensor(
[[0.42470264 0.44720006]
 [0.59241545 0.3188485 ]], shape=(2, 2), dtype=float32)


tf.Tensor(
[[0.13274968 0.00210023 0.34535766 0.42470264 0.44720006]
 [0.887748   0.14301729 0.56532407 0.59241545 0.3188485 ]], shape=(2, 5), dtype=float32)


### **Exercise 1.12**: Create a tensor of shape `(5, 5)` and perform a matrix multiplication with itself. Verify the shape of the resulting tensor.

```python
import tensorflow as tf

tensor = tf.random.uniform((5, 5))
matrix_mul_result = tf.matmul(tensor, tensor)
print(matrix_mul_result)
```

In [ ]:
shape = (5, 5)
tensor = tf.fill(shape, 5)
print(tensor)
print("\n")
result = tf.matmul(tensor, tensor)
print(result)


tf.Tensor(
[[5 5 5 5 5]
 [5 5 5 5 5]
 [5 5 5 5 5]
 [5 5 5 5 5]
 [5 5 5 5 5]], shape=(5, 5), dtype=int32)


tf.Tensor(
[[125 125 125 125 125]
 [125 125 125 125 125]
 [125 125 125 125 125]
 [125 125 125 125 125]
 [125 125 125 125 125]], shape=(5, 5), dtype=int32)


### **Exercise 1.13**: Create a tensor of shape `(4, 4)` and normalize its values to the range `[0, 1]` using min-max normalization.

```python
import tensorflow as tf

tensor = tf.random.uniform((4, 4))
min_val = tf.reduce_min(tensor)
max_val = tf.reduce_max(tensor)
normalized_tensor = (tensor - min_val) / (max_val - min_val)
print(normalized_tensor)
```

In [ ]:
tensor = tf.random.uniform((4, 4), dtype=tf.float32)
print(tensor)
print("\n")
min_val = tf.reduce_min(tensor)
max_val = tf.reduce_max(tensor)
normalized_tensor = (tensor - min_val) / (max_val - min_val)
print(normalized_tensor)

tf.Tensor(
[[0.46605754 0.27148867 0.7850828  0.78346574]
 [0.11966121 0.01524734 0.85850537 0.03094792]
 [0.6826043  0.8229054  0.59567094 0.43822944]
 [0.10708797 0.5360192  0.7849337  0.6473768 ]], shape=(4, 4), dtype=float32)


tf.Tensor(
[[0.53460526 0.3038706  0.9129299  0.91101223]
 [0.12382197 0.         1.         0.01861895]
 [0.79140306 0.95778286 0.6883108  0.5016046 ]
 [0.10891166 0.6175712  0.91275305 0.74962753]], shape=(4, 4), dtype=float32)


### **Exercise 1.14**: Create a TensorFlow tensor with a shape of `(4, 4)` and initialize it with random values between 0 and 1. Slice the tensor to get the bottom-right 2x2 sub-tensor.

 ```python
import tensorflow as tf

tensor = tf.random.uniform((4, 4), minval=0, maxval=1)
sliced_tensor = tensor[2:4, 2:4]|
print(sliced_tensor)
```


In [ ]:
tensor = tf.random.uniform((4,4),minval = 0, maxval = 1, dtype=tf.float32)
print(tensor)
print("\n")
sliced_tensor = tensor[2:4, 2:4]
print(sliced_tensor)

tf.Tensor(
[[0.6969708  0.96683884 0.04934478 0.8124448 ]
 [0.36237144 0.25768936 0.7188425  0.77328897]
 [0.6435652  0.2083683  0.623091   0.7126919 ]
 [0.8834237  0.7987697  0.10878086 0.412269  ]], shape=(4, 4), dtype=float32)


tf.Tensor(
[[0.623091   0.7126919 ]
 [0.10878086 0.412269  ]], shape=(2, 2), dtype=float32)


### **Exercise 1.15**: Create a 1-D tensor with 10 elements using `tf.range`. Reshape this tensor into a 2x5 tensor.

```python
tensor = tf.range(10)
reshaped_tensor = tf.reshape(tensor, (2, 5))
print(reshaped_tensor)
```

In [ ]:
tensor = tf.range(10)
print(tensor)
print("\n")
reshaped_tensor = tf.reshape(tensor, (2, 5))
print(reshaped_tensor)

tf.Tensor([0 1 2 3 4 5 6 7 8 9], shape=(10,), dtype=int32)


tf.Tensor(
[[0 1 2 3 4]
 [5 6 7 8 9]], shape=(2, 5), dtype=int32)


In [ ]:
tensor = tf.range(45)
print(tensor)
print("\n")
reshaped_tensor = tf.reshape(tensor,(3,3,5))
print(reshaped_tensor)

tf.Tensor(
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44], shape=(45,), dtype=int32)


tf.Tensor(
[[[ 0  1  2  3  4]
  [ 5  6  7  8  9]
  [10 11 12 13 14]]

 [[15 16 17 18 19]
  [20 21 22 23 24]
  [25 26 27 28 29]]

 [[30 31 32 33 34]
  [35 36 37 38 39]
  [40 41 42 43 44]]], shape=(3, 3, 5), dtype=int32)


In [ ]:
tensor = tf.range(120)
reshaped_tensor = tf.reshape(tensor,(2,3,4,5))
print(reshaped_tensor)

tf.Tensor(
[[[[  0   1   2   3   4]
   [  5   6   7   8   9]
   [ 10  11  12  13  14]
   [ 15  16  17  18  19]]

  [[ 20  21  22  23  24]
   [ 25  26  27  28  29]
   [ 30  31  32  33  34]
   [ 35  36  37  38  39]]

  [[ 40  41  42  43  44]
   [ 45  46  47  48  49]
   [ 50  51  52  53  54]
   [ 55  56  57  58  59]]]


 [[[ 60  61  62  63  64]
   [ 65  66  67  68  69]
   [ 70  71  72  73  74]
   [ 75  76  77  78  79]]

  [[ 80  81  82  83  84]
   [ 85  86  87  88  89]
   [ 90  91  92  93  94]
   [ 95  96  97  98  99]]

  [[100 101 102 103 104]
   [105 106 107 108 109]
   [110 111 112 113 114]
   [115 116 117 118 119]]]], shape=(2, 3, 4, 5), dtype=int32)


### **Exercise 1.16**: Create a tensor of shape `(5, 5)` filled with ones. Change the center element to 0.

```python
tensor = tf.ones((5, 5))
tensor = tf.tensor_scatter_nd_update(tensor, indices=[[2, 2]], updates=[0])
print(tensor)
```


In [ ]:
tensor = tf.ones((5, 5))
tensor = tf.tensor_scatter_nd_update(tensor, indices=[[2, 2]], updates=[0])
print(tensor)

tf.Tensor(
[[1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]
 [1. 1. 0. 1. 1.]
 [1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1.]], shape=(5, 5), dtype=float32)


### **NOTE:** This alternative method could be used as well.

In [ ]:
tensor = tf.constant([[1, 2], [3, 4], [5, 6]], dtype=tf.int32)

# Indices to update
indices = tf.constant([[0, 0], [1, 1]], dtype=tf.int32)

# New values to update the tensor with
updates = tf.constant([10, 20], dtype=tf.int32)

# Applying the updates
updated_tensor = tf.tensor_scatter_nd_update(tensor, indices, updates)

# Running the session to evaluate the tensor
print(updated_tensor)


tf.Tensor(
[[10  2]
 [ 3 20]
 [ 5  6]], shape=(3, 2), dtype=int32)


### **Exercise 1.17**: Create two 1-D tensors of length 5 with random values. Concatenate them to create a 1-D tensor of length 10.

 ```python
tensor1 = tf.random.uniform((5,))
tensor2 = tf.random.uniform((5,))
concatenated_tensor = tf.concat([tensor1, tensor2], axis=0)
print(concatenated_tensor)
```


In [ ]:
tensor1 = tf.random.uniform((5,))
print(tensor1)
print("\n")
tensor2 = tf.random.uniform((5,))
print(tensor2)
print("\n")
concatenated_tensor = tf.concat([tensor1, tensor2], axis=0)
print(concatenated_tensor)

tf.Tensor([0.36025608 0.20073497 0.6705266  0.8908056  0.44811118], shape=(5,), dtype=float32)


tf.Tensor([0.6351814  0.9134885  0.07533169 0.11312723 0.76014936], shape=(5,), dtype=float32)


tf.Tensor(
[0.36025608 0.20073497 0.6705266  0.8908056  0.44811118 0.6351814
 0.9134885  0.07533169 0.11312723 0.76014936], shape=(10,), dtype=float32)


### **Exercise 1.18**: Create a 3x3 identity matrix using TensorFlow.

```python
identity_matrix = tf.eye(3)
print(identity_matrix)
```

In [ ]:
identity_matrix = tf.eye(3)
print(identity_matrix)

tf.Tensor(
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]], shape=(3, 3), dtype=float32)


### **Exercise 1.19**: Create a tensor of shape `(2, 3)` with random values. Calculate the mean and standard deviation of the tensor.

```python
tensor = tf.random.uniform((2, 3))
mean = tf.reduce_mean(tensor)
stddev = tf.math.reduce_std(tensor)
print("Mean:", mean.numpy())
print("Standard Deviation:", stddev.numpy())
```

In [ ]:
tensor = tf.random.uniform((2,3), dtype=tf.float32)
mean = tf.reduce_mean(tensor)
stddev = tf.math.reduce_std(tensor)
print("Mean:", mean.numpy())
print("Standard Deviation:", stddev.numpy())

Mean: 0.59670806
Standard Deviation: 0.3812475


### **NOTE:** Alternatively, use the following:

In [ ]:
print(mean)
print(stddev)

tf.Tensor(0.59670806, shape=(), dtype=float32)
tf.Tensor(0.3812475, shape=(), dtype=float32)


### **Exercise 1.20**: Create a tensor of shape `(3, 3)` with random values. Normalize the tensor (subtract the mean and divide by the standard deviation).

 ```python
tensor = tf.random.uniform((3, 3))
mean, stddev = tf.reduce_mean(tensor), tf.math.reduce_std(tensor)
normalized_tensor = (tensor - mean) / stddev
print(normalized_tensor)
```

In [ ]:
tensor = tf.random.uniform((3,3), dtype=tf.float32)
print(tensor)
print("\n")
mean, stddev = tf.reduce_mean(tensor), tf.math.reduce_std(tensor)
print("The tensor mean is ", mean.numpy())
print("The tensor stddev is ", stddev.numpy())
print("\n")
normalized_tensor = (tensor - mean) / stddev
print(normalized_tensor)


tf.Tensor(
[[0.5057392  0.187603   0.14445567]
 [0.18466818 0.13067377 0.10724366]
 [0.4967867  0.45638263 0.5123081 ]], shape=(3, 3), dtype=float32)


The tensor mean is  0.30287346
The tensor stddev is  0.17206734


tf.Tensor(
[[ 1.1789904  -0.66991484 -0.9206732 ]
 [-0.686971   -1.0007691  -1.1369374 ]
 [ 1.1269614   0.8921459   1.2171668 ]], shape=(3, 3), dtype=float32)


### **Exercise 1.21**: Create a tensor of shape `(4, 4)` with random values between -1 and 1. Apply the ReLU activation function to this tensor.

```python
tensor = tf.random.uniform((4, 4), minval=-1, maxval=1)
relu_tensor = tf.nn.relu(tensor)
print(relu_tensor)
```

In [ ]:
tensor = tf.random.uniform((4,4), minval=-1, maxval=1, dtype=tf.float32)
print(tensor)
print("\n")
relu_tensor = tf.nn.relu(tensor)
print(relu_tensor)


tf.Tensor(
[[-0.77788234  0.5593064   0.27836728  0.10592794]
 [ 0.3540883  -0.8159628   0.04304075  0.33895922]
 [ 0.678473    0.43741846  0.8848586   0.15097809]
 [ 0.86957574 -0.07371736 -0.7554405   0.635103  ]], shape=(4, 4), dtype=float32)


tf.Tensor(
[[0.         0.5593064  0.27836728 0.10592794]
 [0.3540883  0.         0.04304075 0.33895922]
 [0.678473   0.43741846 0.8848586  0.15097809]
 [0.86957574 0.         0.         0.635103  ]], shape=(4, 4), dtype=float32)


### **Exercise 1.22**: Create a tensor of shape `(5, 5)` with normally distributed random values. Calculate the sum of all elements in the tensor.

```python
tensor = tf.random.normal((5, 5))
sum_tensor = tf.reduce_sum(tensor)
print(sum_tensor)
```

In [ ]:
tensor = tf.random.normal((5, 5))
print(tensor)
print("\n")
sum_tensor = tf.reduce_sum(tensor)
print(sum_tensor)

tf.Tensor(
[[ 0.19211797  1.7210767  -0.67960954 -1.2656827   0.0875546 ]
 [ 1.2149788   1.33337    -0.7580605  -1.6372123  -0.0237071 ]
 [ 1.2195574  -2.3841295   1.9459269   0.7698695  -0.35315275]
 [-0.21510373 -0.56672233 -0.20822129  0.16150315  0.14477125]
 [-0.06870583 -0.7996049   0.9615412  -0.79790604 -0.56094205]], shape=(5, 5), dtype=float32)


tf.Tensor(-0.5664929, shape=(), dtype=float32)


### **Exercise 1.23**: Create two tensors of shape `(2, 2)` with random values. Perform element-wise multiplication and matrix multiplication on these tensors.

```python
tensor1 = tf.random.uniform((2, 2))
tensor2 = tf.random.uniform((2, 2))
elementwise_mul = tf.multiply(tensor1, tensor2)
matrix_mul = tf.matmul(tensor1, tensor2)
print("Element-wise multiplication:\n", elementwise_mul)
print("Matrix multiplication:\n", matrix_mul)
```

In [ ]:
tensor1 = tf.random.uniform((2, 2))
tensor2 = tf.random.uniform((2, 2))
print(tensor1)
print("\n")
print(tensor2)
print("\n")
elementwise_mul = tf.multiply(tensor1, tensor2)
matrix_mul = tf.matmul(tensor1, tensor2)
print("Element-wise multiplication:\n", elementwise_mul)
print("\n")
print("Matrix multiplication:\n", matrix_mul)

tf.Tensor(
[[0.13160324 0.6430135 ]
 [0.5921091  0.552331  ]], shape=(2, 2), dtype=float32)


tf.Tensor(
[[0.72429276 0.98018277]
 [0.33417714 0.74873745]], shape=(2, 2), dtype=float32)


Element-wise multiplication:
 tf.Tensor(
[[0.09531927 0.6302707 ]
 [0.19786932 0.41355088]], shape=(2, 2), dtype=float32)


Matrix multiplication:
 tf.Tensor(
[[0.31019968 0.6104435 ]
 [0.6134367  0.99392605]], shape=(2, 2), dtype=float32)


### **Exercise 1.24**: Create a tensor of shape `(3, 4)` with random values. Transpose this tensor.

```python
tensor = tf.random.uniform((3, 4))
transposed_tensor = tf.transpose(tensor)
print(transposed_tensor)
```

In [ ]:
tensor = tf.random.uniform((3, 4))
print(tensor)
print("\n")
transposed_tensor = tf.transpose(tensor)
print(transposed_tensor)

tf.Tensor(
[[0.06667948 0.5197476  0.07745123 0.6319295 ]
 [0.08938646 0.71627486 0.9679235  0.12232053]
 [0.4180194  0.64101136 0.57963717 0.35457957]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[0.06667948 0.08938646 0.4180194 ]
 [0.5197476  0.71627486 0.64101136]
 [0.07745123 0.9679235  0.57963717]
 [0.6319295  0.12232053 0.35457957]], shape=(4, 3), dtype=float32)


### **Exercise 1.25**: Create a tensor of shape `(2, 3, 4)` with random values. Reshape this tensor into a 2-D tensor of shape `(6, 4)`.

```python
tensor = tf.random.uniform((2, 3, 4))
reshaped_tensor = tf.reshape(tensor, (6, 4))
print(reshaped_tensor)
```


In [ ]:
tensor = tf.random.uniform((2, 3, 4))
print(tensor)
print("\n")
reshaped_tensor = tf.reshape(tensor, (6, 4))
print(reshaped_tensor)

tf.Tensor(
[[[0.7658713  0.58283174 0.07502723 0.46517015]
  [0.6806258  0.51526904 0.49553466 0.28917456]
  [0.5618943  0.8792777  0.45804846 0.6958808 ]]

 [[0.65646803 0.8502265  0.48747313 0.68744373]
  [0.12465739 0.84266555 0.535537   0.0020268 ]
  [0.25344813 0.08947051 0.17269301 0.9221531 ]]], shape=(2, 3, 4), dtype=float32)


tf.Tensor(
[[0.7658713  0.58283174 0.07502723 0.46517015]
 [0.6806258  0.51526904 0.49553466 0.28917456]
 [0.5618943  0.8792777  0.45804846 0.6958808 ]
 [0.65646803 0.8502265  0.48747313 0.68744373]
 [0.12465739 0.84266555 0.535537   0.0020268 ]
 [0.25344813 0.08947051 0.17269301 0.9221531 ]], shape=(6, 4), dtype=float32)


In [ ]:
tensor = tf.random.uniform((3, 4, 2))
print(tensor)
print("\n")
reshaped_tensor = tf.reshape(tensor, (6, 4))
print(reshaped_tensor)

tf.Tensor(
[[[0.94488597 0.5145209 ]
  [0.7138705  0.27301633]
  [0.9180417  0.67404115]
  [0.52818286 0.5724189 ]]

 [[0.39344656 0.42576134]
  [0.814363   0.58020055]
  [0.33641386 0.3047248 ]
  [0.67971075 0.03682518]]

 [[0.28652215 0.16238546]
  [0.4463042  0.11843526]
  [0.375818   0.04864728]
  [0.85832083 0.15718138]]], shape=(3, 4, 2), dtype=float32)


tf.Tensor(
[[0.94488597 0.5145209  0.7138705  0.27301633]
 [0.9180417  0.67404115 0.52818286 0.5724189 ]
 [0.39344656 0.42576134 0.814363   0.58020055]
 [0.33641386 0.3047248  0.67971075 0.03682518]
 [0.28652215 0.16238546 0.4463042  0.11843526]
 [0.375818   0.04864728 0.85832083 0.15718138]], shape=(6, 4), dtype=float32)


In [ ]:
tensor = tf.random.uniform((3, 4, 3))
print(tensor)
print("\n")
reshaped_tensor = tf.reshape(tensor, (9, 4))
print(reshaped_tensor)

tf.Tensor(
[[[0.9358349  0.70090413 0.658484  ]
  [0.04049337 0.57098496 0.01053798]
  [0.42961192 0.25810826 0.22988975]
  [0.9329432  0.43180108 0.91119885]]

 [[0.59612453 0.11149096 0.60991776]
  [0.4798348  0.50438464 0.23230016]
  [0.8554419  0.54522395 0.9565302 ]
  [0.75052536 0.2502402  0.1482917 ]]

 [[0.90709674 0.03364861 0.71141064]
  [0.77366567 0.92848575 0.5938349 ]
  [0.55807257 0.3462646  0.25349164]
  [0.39053977 0.00560987 0.4848665 ]]], shape=(3, 4, 3), dtype=float32)


tf.Tensor(
[[0.9358349  0.70090413 0.658484   0.04049337]
 [0.57098496 0.01053798 0.42961192 0.25810826]
 [0.22988975 0.9329432  0.43180108 0.91119885]
 [0.59612453 0.11149096 0.60991776 0.4798348 ]
 [0.50438464 0.23230016 0.8554419  0.54522395]
 [0.9565302  0.75052536 0.2502402  0.1482917 ]
 [0.90709674 0.03364861 0.71141064 0.77366567]
 [0.92848575 0.5938349  0.55807257 0.3462646 ]
 [0.25349164 0.39053977 0.00560987 0.4848665 ]], shape=(9, 4), dtype=float32)


In [ ]:
tensor = tf.random.uniform((9, 4))
print(tensor)
print("\n")
reshaped_tensor = tf.reshape(tensor, (3,4,3))
print(reshaped_tensor)

tf.Tensor(
[[0.15457284 0.18791449 0.43169713 0.80818236]
 [0.6828054  0.8385073  0.5317954  0.03671932]
 [0.14129424 0.4076743  0.7442496  0.9805175 ]
 [0.8302387  0.2814039  0.1721232  0.42819917]
 [0.49002385 0.47037518 0.74634886 0.256796  ]
 [0.6543592  0.09454763 0.412403   0.44560742]
 [0.08534276 0.39247918 0.50008416 0.09329247]
 [0.750479   0.20401478 0.66917515 0.40154672]
 [0.63235784 0.14520216 0.32950735 0.24789393]], shape=(9, 4), dtype=float32)


tf.Tensor(
[[[0.15457284 0.18791449 0.43169713]
  [0.80818236 0.6828054  0.8385073 ]
  [0.5317954  0.03671932 0.14129424]
  [0.4076743  0.7442496  0.9805175 ]]

 [[0.8302387  0.2814039  0.1721232 ]
  [0.42819917 0.49002385 0.47037518]
  [0.74634886 0.256796   0.6543592 ]
  [0.09454763 0.412403   0.44560742]]

 [[0.08534276 0.39247918 0.50008416]
  [0.09329247 0.750479   0.20401478]
  [0.66917515 0.40154672 0.63235784]
  [0.14520216 0.32950735 0.24789393]]], shape=(3, 4, 3), dtype=float32)


### **Exercise 1.26**: Create a tensor of shape `(3, 3)` with random values. Extract the diagonal elements of this tensor.

```python
tensor = tf.random.uniform((3, 3))
diagonal = tf.linalg.diag_part(tensor)
print(diagonal)
```

In [ ]:
tensor = tf.random.uniform((3, 3))
print(tensor)
print("\n")
diagonal = tf.linalg.diag_part(tensor)
print(diagonal)

tf.Tensor(
[[0.9826915  0.6717384  0.06726861]
 [0.01254535 0.22600806 0.80312145]
 [0.06481028 0.8579792  0.73611903]], shape=(3, 3), dtype=float32)


tf.Tensor([0.9826915  0.22600806 0.73611903], shape=(3,), dtype=float32)


### **Exercise 1.27**: Create a tensor of shape `(4, 4)` with values from 1 to 16. Reverse the order of elements in each row.

```python
tensor = tf.reshape(tf.range(1, 17), (4, 4))
reversed_tensor = tf.reverse(tensor, axis=[1])
print(reversed_tensor)
```

In [ ]:
tensor = tf.reshape(tf.range(1, 17), (4, 4))
print(tensor)
print("\n")
reversed_tensor = tf.reverse(tensor, axis=[1])
print(reversed_tensor)

tf.Tensor(
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]], shape=(4, 4), dtype=int32)


tf.Tensor(
[[ 4  3  2  1]
 [ 8  7  6  5]
 [12 11 10  9]
 [16 15 14 13]], shape=(4, 4), dtype=int32)


### ### **Exercise 1.28**: Create a tensor of shape `(4, 4)` with values from 1 to 16. Reverse the order of elements in each column.

```python
tensor = tf.reshape(tf.range(1, 17), (4, 4))
reversed_tensor = tf.reverse(tensor, axis=[0])
print(reversed_tensor)
```

In [ ]:
tensor = tf.reshape(tf.range(1, 17), (4, 4))
print(tensor)
print("\n")
reversed_tensor = tf.reverse(tensor, axis=[0])
print(reversed_tensor)

tf.Tensor(
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]], shape=(4, 4), dtype=int32)


tf.Tensor(
[[13 14 15 16]
 [ 9 10 11 12]
 [ 5  6  7  8]
 [ 1  2  3  4]], shape=(4, 4), dtype=int32)


In [ ]:
tensor = tf.reshape(tf.range(1, 25), (2,3,4))
print(tensor)
print("\n")
reversed_tensor = tf.reverse(tensor, axis=[0])
print(reversed_tensor)
print("\n")
reversed_tensor = tf.reverse(tensor, axis=[1])
print(reversed_tensor)
print("\n")
reversed_tensor = tf.reverse(tensor, axis=[2])
print(reversed_tensor)

tf.Tensor(
[[[ 1  2  3  4]
  [ 5  6  7  8]
  [ 9 10 11 12]]

 [[13 14 15 16]
  [17 18 19 20]
  [21 22 23 24]]], shape=(2, 3, 4), dtype=int32)


tf.Tensor(
[[[13 14 15 16]
  [17 18 19 20]
  [21 22 23 24]]

 [[ 1  2  3  4]
  [ 5  6  7  8]
  [ 9 10 11 12]]], shape=(2, 3, 4), dtype=int32)


tf.Tensor(
[[[ 9 10 11 12]
  [ 5  6  7  8]
  [ 1  2  3  4]]

 [[21 22 23 24]
  [17 18 19 20]
  [13 14 15 16]]], shape=(2, 3, 4), dtype=int32)


tf.Tensor(
[[[ 4  3  2  1]
  [ 8  7  6  5]
  [12 11 10  9]]

 [[16 15 14 13]
  [20 19 18 17]
  [24 23 22 21]]], shape=(2, 3, 4), dtype=int32)


### **Exercise 1.29**: Create a tensor of shape `(5, 5)` with random values. Calculate the determinant of this tensor.

```python
tensor = tf.random.uniform((5, 5))
determinant = tf.linalg.det(tensor)
print(determinant)
```

In [ ]:
tensor = tf.random.uniform((5, 5))
determinant = tf.linalg.det(tensor)
print(determinant)

tf.Tensor(0.066316605, shape=(), dtype=float32)


### **Exercise 1.30**: Create a tensor of shape `(3, 3)` with random values. Calculate the inverse of this tensor (if it's invertible).

```python
tensor = tf.random.uniform((3, 3))
inverse_tensor = tf.linalg.inv(tensor)
print(inverse_tensor)
```

In [ ]:
tensor = tf.random.uniform((3, 3))
print(tensor)
print("\n")
inverse_tensor = tf.linalg.inv(tensor)
print(inverse_tensor)

tf.Tensor(
[[0.6977099  0.02439511 0.9139519 ]
 [0.02176034 0.8284366  0.37790084]
 [0.85830176 0.59788954 0.28219283]], shape=(3, 3), dtype=float32)


tf.Tensor(
[[-0.01254268 -0.8636476   1.1971831 ]
 [-0.5093488   0.9404785   0.39020383]
 [ 1.1173201   0.6342045  -0.92434365]], shape=(3, 3), dtype=float32)


### **Exercise 1.31**: Create a tensor of shape `(4, 4)` with values from 1 to 16. Flatten this tensor into a 1-D tensor.

```python
tensor = tf.reshape(tf.range(1, 17), (4, 4))
flattened_tensor = tf.reshape(tensor, [-1])
print(flattened_tensor)
```

In [ ]:
tensor = tf.reshape(tf.range(1, 17), (4, 4))
print(tensor)
print("\n")
flattened_tensor = tf.reshape(tensor, [-1])
print(flattened_tensor)

tf.Tensor(
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]], shape=(4, 4), dtype=int32)


tf.Tensor([ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16], shape=(16,), dtype=int32)


###  **Exercise 1.32**: Create a tensor of shape `(3, 3, 3)` with random values. Slice this tensor to get the first two elements along the first dimension.

```python
tensor = tf.random.uniform((3, 3, 3))
sliced_tensor = tensor[:2, :, :]
print(sliced_tensor)
```

In [ ]:
tensor = tf.random.uniform((3, 3, 3))
print(tensor)
print("\n")
sliced_tensor0 = tensor[:2, :, :]
print(sliced_tensor)
sliced_tensor1 = tensor[:2, :2, :]
print(sliced_tensor)
sliced_tensor2 = tensor[:2, :2, :2]
print(sliced_tensor)

tf.Tensor(
[[[0.726774   0.09455466 0.98843205]
  [0.441522   0.7863184  0.58635056]
  [0.9326335  0.7546983  0.9889581 ]]

 [[0.577229   0.43325675 0.7767762 ]
  [0.5943233  0.45460105 0.9382982 ]
  [0.74622834 0.429101   0.22449958]]

 [[0.32361543 0.76701343 0.8127787 ]
  [0.34617305 0.29057884 0.09946346]
  [0.15137768 0.5925269  0.06499994]]], shape=(3, 3, 3), dtype=float32)


tf.Tensor(
[[[0.2928983  0.13033259]
  [0.4118985  0.5480361 ]]

 [[0.35642326 0.05712414]
  [0.36564064 0.16358149]]], shape=(2, 2, 2), dtype=float32)
tf.Tensor(
[[[0.2928983  0.13033259]
  [0.4118985  0.5480361 ]]

 [[0.35642326 0.05712414]
  [0.36564064 0.16358149]]], shape=(2, 2, 2), dtype=float32)
tf.Tensor(
[[[0.2928983  0.13033259]
  [0.4118985  0.5480361 ]]

 [[0.35642326 0.05712414]
  [0.36564064 0.16358149]]], shape=(2, 2, 2), dtype=float32)


In [ ]:
tensor = tf.random.uniform((3, 3, 3))
print(tensor)
print("\n")
sliced_tensor = tensor[1:2, :1, :1]
print(sliced_tensor)


tf.Tensor(
[[[0.29438627 0.99649215 0.88832223]
  [0.5622302  0.9798237  0.34900773]
  [0.09645367 0.04779387 0.8321619 ]]

 [[0.1753596  0.94031584 0.63521624]
  [0.00839889 0.73268056 0.26641762]
  [0.08868098 0.28122413 0.7277858 ]]

 [[0.57776606 0.33785057 0.3724817 ]
  [0.62784994 0.8479384  0.8305588 ]
  [0.6040845  0.8737924  0.13997447]]], shape=(3, 3, 3), dtype=float32)


tf.Tensor([[[0.1753596]]], shape=(1, 1, 1), dtype=float32)


### **Exercise 1.33**: Create a tensor of shape `(6, 6)` with random values. Select and print every other row and column.

```python
tensor = tf.random.uniform((6, 6))
sliced_tensor = tensor[::2, ::2]
print(sliced_tensor)
```

In [ ]:
tensor = tf.random.uniform((9, 4))
print(tensor)
print("\n")
sliced_tensor = tensor[::3, ::3]
print(sliced_tensor)

tf.Tensor(
[[0.2876389  0.2838689  0.24167025 0.2392205 ]
 [0.1155535  0.6878636  0.22255778 0.26586723]
 [0.6652063  0.57510495 0.61145675 0.82786083]
 [0.75984156 0.5787777  0.8690994  0.11727417]
 [0.86857927 0.5910324  0.332489   0.4820521 ]
 [0.37193477 0.7416289  0.34761012 0.83183634]
 [0.18892622 0.8688972  0.15095377 0.76019895]
 [0.13809705 0.04393101 0.8479638  0.48946047]
 [0.99992085 0.09968555 0.2684431  0.6949419 ]], shape=(9, 4), dtype=float32)


tf.Tensor(
[[0.2876389  0.2392205 ]
 [0.75984156 0.11727417]
 [0.18892622 0.76019895]], shape=(3, 2), dtype=float32)


### **Exercise 1.34**: Create a tensor of shape `(3, 3, 3)` with random values. Sum the tensor along the second dimension.

```python
tensor = tf.random.uniform((3, 3, 3))
summed_tensor = tf.reduce_sum(tensor, axis=1)
print(summed_tensor)
```

In [ ]:
tensor = tf.random.uniform((3,3,3))
print(tensor)
print("\n")
summed_tensor = tf.reduce_sum(tensor, axis=1)
print(summed_tensor)

tf.Tensor(
[[[0.35960162 0.6449121  0.03177512]
  [0.46776795 0.9073645  0.5707902 ]
  [0.3714559  0.4989164  0.8294594 ]]

 [[0.36843622 0.19478047 0.00617957]
  [0.57095325 0.89587224 0.93769217]
  [0.13420367 0.28416562 0.77790284]]

 [[0.13094127 0.8727237  0.8160578 ]
  [0.4524815  0.6777934  0.02721226]
  [0.05320191 0.41025496 0.0081836 ]]], shape=(3, 3, 3), dtype=float32)


tf.Tensor(
[[1.1988255  2.051193   1.4320247 ]
 [1.0735931  1.3748183  1.7217746 ]
 [0.6366247  1.960772   0.85145366]], shape=(3, 3), dtype=float32)


tf.Tensor(
[[0.8589791 1.7124163 0.8540125]
 [1.4912027 2.48103   1.5356946]
 [0.5588615 1.193337  1.6155459]], shape=(3, 3), dtype=float32)


tf.Tensor(
[[1.0362889  1.9459226  1.6998317 ]
 [0.56939626 2.4045177  1.1962721 ]
 [1.8197228  1.1574872  0.47164047]], shape=(3, 3), dtype=float32)


In [ ]:
tensor = tf.reshape(tf.range(0.1, 2.5, 0.1, dtype=tf.float32), (2, 3, 4))
tensor = tf.round(tensor * 10) / 10
print(tensor)
print("\n")
summed_tensor = tf.reduce_sum(tensor, axis=0)
print(summed_tensor)
print("\n")
summed_tensor = tf.reduce_sum(tensor, axis=1)
print(summed_tensor)
print("\n")
summed_tensor = tf.reduce_sum(tensor, axis=2)
print(summed_tensor)

tf.Tensor(
[[[0.1 0.2 0.3 0.4]
  [0.5 0.6 0.7 0.8]
  [0.9 1.  1.1 1.2]]

 [[1.3 1.4 1.5 1.6]
  [1.7 1.8 1.9 2. ]
  [2.1 2.2 2.3 2.4]]], shape=(2, 3, 4), dtype=float32)


tf.Tensor(
[[1.4       1.6       1.8       2.       ]
 [2.2       2.4       2.6       2.8      ]
 [3.        3.2       3.4       3.6000001]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[1.5       1.8       2.1       2.4      ]
 [5.1       5.3999996 5.7       6.       ]], shape=(2, 4), dtype=float32)


tf.Tensor(
[[1.        2.6       4.2      ]
 [5.7999997 7.4       9.       ]], shape=(2, 3), dtype=float32)


### **Exercise 1.35**: Create a tensor of shape `(4, 4)` with random values. Split this tensor into four 2x2 sub-tensors.

```python
tensor = tf.random.uniform((4, 4))
sub_tensors = tf.split(tensor, num_or_size_splits=2, axis=0)
sub_tensors = [tf.split(sub_tensor, num_or_size_splits=2, axis=1) for sub_tensor in sub_tensors]
print(sub_tensors[0][0])
print(sub_tensors[0][1])
print(sub_tensors[1][0])
print(sub_tensors[1][1])
```

In [ ]:
tensor = tf.random.uniform((4, 4))
print(tensor)
print("\n")
sub_tensors = tf.split(tensor, num_or_size_splits=2, axis=0)
print(sub_tensors)
print("\n")
sub_tensors = [tf.split(sub_tensor, num_or_size_splits=2, axis=1) for sub_tensor in sub_tensors]
print(sub_tensors)
print("\n")
print(sub_tensors[0][0])
print("\n")
print(sub_tensors[0][1])
print("\n")
print(sub_tensors[1][0])
print("\n")
print(sub_tensors[1][1])

tf.Tensor(
[[0.48285758 0.26693928 0.18382025 0.2302196 ]
 [0.3255893  0.5042473  0.63711214 0.95684874]
 [0.35219467 0.00682306 0.2579291  0.75935197]
 [0.503266   0.39050043 0.81763744 0.30535412]], shape=(4, 4), dtype=float32)


[<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0.48285758, 0.26693928, 0.18382025, 0.2302196 ],
       [0.3255893 , 0.5042473 , 0.63711214, 0.95684874]], dtype=float32)>, <tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0.35219467, 0.00682306, 0.2579291 , 0.75935197],
       [0.503266  , 0.39050043, 0.81763744, 0.30535412]], dtype=float32)>]


[[<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[0.48285758, 0.26693928],
       [0.3255893 , 0.5042473 ]], dtype=float32)>, <tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[0.18382025, 0.2302196 ],
       [0.63711214, 0.95684874]], dtype=float32)>], [<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[0.35219467, 0.00682306],
       [0.503266  , 0.39050043]], dtype=float32)>, <tf.T

### **Exercise 1.36**: Create a tensor of shape `(3, 4)` with random values. Stack this tensor with itself along a new dimension to create a tensor of shape `(2, 3, 4)`.

```python
tensor = tf.random.uniform((3, 4))
stacked_tensor = tf.stack([tensor, tensor], axis=0)
print(stacked_tensor)
```

In [ ]:
tensor = tf.random.uniform((3, 4))
print(tensor)
print("\n")
stacked_tensor = tf.stack([tensor, tensor], axis=0)
print(stacked_tensor)

tf.Tensor(
[[0.67470443 0.69526803 0.06844628 0.9604924 ]
 [0.9215143  0.23192132 0.42629695 0.16379893]
 [0.28489304 0.19684625 0.6709516  0.6633531 ]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[[0.67470443 0.69526803 0.06844628 0.9604924 ]
  [0.9215143  0.23192132 0.42629695 0.16379893]
  [0.28489304 0.19684625 0.6709516  0.6633531 ]]

 [[0.67470443 0.69526803 0.06844628 0.9604924 ]
  [0.9215143  0.23192132 0.42629695 0.16379893]
  [0.28489304 0.19684625 0.6709516  0.6633531 ]]], shape=(2, 3, 4), dtype=float32)


### **Exercise 1.37**: Create two tensors of shape `(3, 3)` with random values. Compute the element-wise maximum and minimum of these tensors.

```python
tensor1 = tf.random.uniform((3, 3))
tensor2 = tf.random.uniform((3, 3))
max_tensor = tf.maximum(tensor1, tensor2)
min_tensor = tf.minimum(tensor1, tensor2)
print("Element-wise maximum:\n", max_tensor)
print("Element-wise minimum:\n", min_tensor)
```

In [ ]:
tensor1 = tf.random.uniform((3, 3))
print(tensor1)
print("\n")
tensor2 = tf.random.uniform((3, 3))
print(tensor2)
print("\n")
max_tensor = tf.maximum(tensor1, tensor2)
min_tensor = tf.minimum(tensor1, tensor2)
print("Element-wise maximum:\n", max_tensor)
print("\n")
print("Element-wise minimum:\n", min_tensor)

tf.Tensor(
[[0.04978693 0.17020154 0.7548919 ]
 [0.09679735 0.5094514  0.04136527]
 [0.5542927  0.14712596 0.31040072]], shape=(3, 3), dtype=float32)


tf.Tensor(
[[0.7081362  0.69033563 0.21733165]
 [0.7666391  0.6396003  0.48016143]
 [0.9269599  0.5956979  0.44160295]], shape=(3, 3), dtype=float32)


Element-wise maximum:
 tf.Tensor(
[[0.7081362  0.69033563 0.7548919 ]
 [0.7666391  0.6396003  0.48016143]
 [0.9269599  0.5956979  0.44160295]], shape=(3, 3), dtype=float32)


Element-wise minimum:
 tf.Tensor(
[[0.04978693 0.17020154 0.21733165]
 [0.09679735 0.5094514  0.04136527]
 [0.5542927  0.14712596 0.31040072]], shape=(3, 3), dtype=float32)


### **Exercise 1.38**: Create a tensor of shape `(4, 4)` with random values. Clip the values in this tensor to be between 0.2 and 0.8.

```python
tensor = tf.random.uniform((4, 4))
clipped_tensor = tf.clip_by_value(tensor, clip_value_min=0.2, clip_value_max=0.8)
print(clipped_tensor)
```

In [ ]:
tensor = tf.random.uniform((4, 4))
print(tensor)
print("\n")
clipped_tensor = tf.clip_by_value(tensor, clip_value_min=0.2, clip_value_max=0.8)
print(clipped_tensor)

tf.Tensor(
[[0.6479721  0.5655178  0.31252658 0.690218  ]
 [0.3194183  0.8105973  0.2746669  0.03204834]
 [0.6319951  0.7076889  0.19439995 0.5601405 ]
 [0.6642617  0.49075532 0.67101884 0.27476668]], shape=(4, 4), dtype=float32)


tf.Tensor(
[[0.6479721  0.5655178  0.31252658 0.690218  ]
 [0.3194183  0.8        0.2746669  0.2       ]
 [0.6319951  0.7076889  0.2        0.5601405 ]
 [0.6642617  0.49075532 0.67101884 0.27476668]], shape=(4, 4), dtype=float32)


### **Exercise 1.39**: Create two tensors of shape `(3, 4)` with random values. Perform matrix multiplication of these tensors with another tensor of shape `(4, 5)` using broadcasting.

```python
tensor1 = tf.random.uniform((2, 3, 4))
tensor2 = tf.random.uniform((4, 5))

# Matrix multiplication with broadcasting
result = tf.matmul(tensor2, tensor2)
print(result)
```

In [ ]:
tensor1 = tf.random.uniform((2, 3, 4))
print(tensor1)
print("\n")
tensor2 = tf.random.uniform((4, 5))
print(tensor2)
print("\n")

# Matrix multiplication with broadcasting
result = tf.matmul(tensor1, tensor2)
print(result)

tf.Tensor(
[[[0.8780794  0.26239955 0.07595193 0.40588307]
  [0.86252415 0.533973   0.62635756 0.18634748]
  [0.04382479 0.0490998  0.43202293 0.23137295]]

 [[0.44416046 0.27652586 0.9120767  0.25665843]
  [0.05764306 0.07638669 0.5451591  0.79119563]
  [0.81265295 0.5181327  0.6603806  0.5202035 ]]], shape=(2, 3, 4), dtype=float32)


tf.Tensor(
[[0.9000511  0.3208164  0.90875804 0.08648014 0.7146832 ]
 [0.6287576  0.35423315 0.15925288 0.63336766 0.19517016]
 [0.55595756 0.34833002 0.33700645 0.8003528  0.21154821]
 [0.06543839 0.8623445  0.54180574 0.09577477 0.40850246]], shape=(4, 5), dtype=float32)


tf.Tensor(
[[[1.0240885  0.75112027 1.0852556  0.34179354 0.8606329 ]
  [1.4724778  0.8447377  1.1809132  0.9319468  0.92927533]
  [0.3256435  0.38146225 0.3185991  0.40281868 0.22681376]]

 [[1.0975062  0.77948034 0.8941067  0.9681182  0.6691974 ]
  [0.4547703  0.91773003 0.67694485 0.565462   0.49463773]
  [1.4583939  1.1228771  1.3254207  0.9768067  1.0341202 ]]], shape=(2, 3, 5),

### **Exercise 1.40**: Create a tensor of shape `(4, 4, 4)` with random values. Transpose this tensor, swapping the second and third dimensions.

```python
tensor = tf.random.uniform((2, 3, 4, 5))
transposed = tf.transpose(tensor, perm=[0, 2, 1, 3])
print(transposed)
```

In [ ]:
tensor = tf.random.uniform((2, 3, 4, 5))
print(tensor)
print("\n")
transposed = tf.transpose(tensor, perm=[0, 2, 1, 3])
print(transposed)

tf.Tensor(
[[[[0.32446933 0.62303805 0.98415077 0.2719053  0.58079517]
   [0.5853435  0.4870553  0.29575384 0.5928421  0.12469459]
   [0.04310882 0.39283288 0.7261604  0.3441602  0.3404374 ]
   [0.17259204 0.4941944  0.02754986 0.8631463  0.8216777 ]]

  [[0.88642037 0.26471853 0.4689802  0.9818176  0.8838476 ]
   [0.8032005  0.01216292 0.8773105  0.66653407 0.0692178 ]
   [0.62811196 0.69068503 0.00588036 0.8535117  0.62828374]
   [0.5034963  0.80665123 0.01096976 0.88732433 0.42610538]]

  [[0.42652893 0.17313147 0.7076216  0.6324519  0.19139922]
   [0.12684536 0.31938648 0.47498226 0.5302758  0.8931258 ]
   [0.9004781  0.657133   0.9880712  0.3665681  0.75365937]
   [0.08968234 0.6970904  0.39361703 0.12195802 0.55396867]]]


 [[[0.69821656 0.7602478  0.33692968 0.45059443 0.75917244]
   [0.67404306 0.04071832 0.04887354 0.68329525 0.67216814]
   [0.65858316 0.47917008 0.16302168 0.04102874 0.9000932 ]
   [0.9255756  0.5997509  0.74041915 0.5545734  0.275913  ]]

  [[0.48918438 0.29

### **Exercise 1.41**: Create two tensors of shape `(3, 4)` with random values. Compute the element-wise maximum and minimum of these tensors.

```python
tensor1 = tf.random.uniform((3, 4))
tensor2 = tf.random.uniform((3, 4))
print(tensor1)
print("\n")
print(tensor2)
print("\n")
maximum = tf.maximum(tensor1, tensor2)
print(maximum)
print("\n")
minimum = tf.minimum(tensor1, tensor2)
print(minimum)
```


In [ ]:
tensor1 = tf.random.uniform((3, 4))
tensor2 = tf.random.uniform((3, 4))
print(tensor1)
print("\n")
print(tensor2)
print("\n")
maximum = tf.maximum(tensor1, tensor2)
print(maximum)
print("\n")
minimum = tf.minimum(tensor1, tensor2)
print(minimum)

tf.Tensor(
[[0.61362934 0.12119591 0.2735225  0.00255346]
 [0.3645153  0.67077374 0.5799446  0.593295  ]
 [0.20093668 0.5040746  0.18972886 0.8169426 ]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[0.17241347 0.7937285  0.30165267 0.4567914 ]
 [0.770964   0.22982252 0.68450284 0.6977658 ]
 [0.47637033 0.84248996 0.4314127  0.39556432]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[0.61362934 0.7937285  0.30165267 0.4567914 ]
 [0.770964   0.67077374 0.68450284 0.6977658 ]
 [0.47637033 0.84248996 0.4314127  0.8169426 ]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[0.17241347 0.12119591 0.2735225  0.00255346]
 [0.3645153  0.22982252 0.5799446  0.593295  ]
 [0.20093668 0.5040746  0.18972886 0.39556432]], shape=(3, 4), dtype=float32)


### **Exercise 1.42**: Create a tensor of shape `(5, 5)` with random values. Slice this tensor to get a `(2, 2)` sub-tensor from the center and assign new values to this sliced part.

```python
tensor = tf.Variable(tf.random.uniform((5, 5)))

# Slice and assign
tensor[1:3, 2:4].assign(tf.ones((2, 2)))

print(tensor)
```

In [ ]:
tensor = tf.Variable(tf.random.uniform((5, 5)))
print(tensor)
print("\n")
# Slice and assign
tensor[1:3, 2:4].assign(tf.ones((2, 2)))
print(tensor)

<tf.Variable 'Variable:0' shape=(5, 5) dtype=float32, numpy=
array([[0.4325062 , 0.28824604, 0.90831804, 0.09905183, 0.22147405],
       [0.46390414, 0.7664652 , 0.9632522 , 0.2792846 , 0.9182683 ],
       [0.21588087, 0.7538049 , 0.88359547, 0.64630425, 0.9303011 ],
       [0.06127071, 0.97695935, 0.21008551, 0.11843157, 0.7989433 ],
       [0.90570736, 0.07175899, 0.05108011, 0.98714995, 0.46430957]],
      dtype=float32)>


<tf.Variable 'Variable:0' shape=(5, 5) dtype=float32, numpy=
array([[0.4325062 , 0.28824604, 0.90831804, 0.09905183, 0.22147405],
       [0.46390414, 0.7664652 , 1.        , 1.        , 0.9182683 ],
       [0.21588087, 0.7538049 , 1.        , 1.        , 0.9303011 ],
       [0.06127071, 0.97695935, 0.21008551, 0.11843157, 0.7989433 ],
       [0.90570736, 0.07175899, 0.05108011, 0.98714995, 0.46430957]],
      dtype=float32)>


### **Exercise 1.43**: Define a function \( z = x^2 + y^2 \). Create two variables `x` and `y` initialized with random values. Compute the gradient of \( z \) with respect to `x` and `y`.

```python
x = tf.Variable(3.0)
y = tf.Variable(4.0)

# Function
with tf.GradientTape() as tape:
    z = x**2 + y**2

# Compute gradient
gradients = tape.gradient(z, [x, y])

print(gradients)
```

In [ ]:
x = tf.Variable(3.0)
y = tf.Variable(4.0)

# Function
with tf.GradientTape() as tape:
    z = x**2 + y**2

# Compute gradient
gradients = tape.gradient(z, [x, y])

print(gradients)

[<tf.Tensor: shape=(), dtype=float32, numpy=6.0>, <tf.Tensor: shape=(), dtype=float32, numpy=8.0>]


### **Exercise 1.44**: Create a tensor of shape `(4, 4, 4)` with random values. Reshape this tensor to have a shape `(8, -1)`, inferring the second dimension.

```python
tensor = tf.random.uniform((4, 4, 4))

# Reshape with inferred dimension
reshaped = tf.reshape(tensor, (-1, 8))

print(reshaped)
```

In [ ]:
tensor = tf.random.uniform((4, 4, 4))
print(tensor)
print("\n")
# Reshape with inferred dimension
reshaped = tf.reshape(tensor, (-1, 8))
print(reshaped)

tf.Tensor(
[[[0.39202952 0.12933671 0.89229774 0.21568716]
  [0.5189315  0.54685616 0.91969323 0.1384418 ]
  [0.7387233  0.6877178  0.01725447 0.29777896]
  [0.5033865  0.4073428  0.12930667 0.47961998]]

 [[0.36173677 0.19173634 0.8928641  0.52175117]
  [0.9752146  0.9304764  0.03006053 0.50270665]
  [0.87035894 0.85412574 0.55552185 0.7563096 ]
  [0.98305357 0.5632361  0.9541104  0.98903096]]

 [[0.30577564 0.4996208  0.55603814 0.50444496]
  [0.7078717  0.7890576  0.20206034 0.72639   ]
  [0.49509478 0.8939104  0.7782403  0.33042037]
  [0.03286135 0.9814327  0.41533935 0.06160271]]

 [[0.1498574  0.52639174 0.5000609  0.8085041 ]
  [0.3234067  0.10109699 0.63134885 0.7816037 ]
  [0.93181586 0.55360866 0.10659075 0.7475195 ]
  [0.6670209  0.54243815 0.7739105  0.5259634 ]]], shape=(4, 4, 4), dtype=float32)


tf.Tensor(
[[0.39202952 0.12933671 0.89229774 0.21568716 0.5189315  0.54685616
  0.91969323 0.1384418 ]
 [0.7387233  0.6877178  0.01725447 0.29777896 0.5033865  0.4073428
  0.129

### **Exercise 1.45**: Create a tensor of integer indices `[0, 1, 2, 1]`. Convert these indices to a one-hot encoded tensor with a depth of 4.

```python
indices = tf.constant([0, 1, 2, 1])

# One-hot encoding
one_hot = tf.one_hot(indices, depth=3)

print(one_hot)
```

In [ ]:
indices = tf.constant([0, 1, 2, 1])
print(indices)
print("\n")
# One-hot encoding
one_hot = tf.one_hot(indices, depth=3)
print(one_hot)

tf.Tensor([0 1 2 1], shape=(4,), dtype=int32)


tf.Tensor(
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]
 [0. 1. 0.]], shape=(4, 3), dtype=float32)


### **Exercise 1.46**: Create a batch of 3 matrices, each of shape `(4, 4)`, with random values. Compute the inverse of each matrix in the batch.

```python
# Batch of matrices
batch_matrices = tf.random.uniform((3, 4, 4))

# Batch matrix inversion
inverted_matrices = tf.linalg.inv(batch_matrices)

print(inverted_matrices)
```

In [ ]:
batch_matrices = tf.random.uniform((3, 4, 4))
print(batch_matrices)
print("\n")
# Batch matrix inversion
inverted_matrices = tf.linalg.inv(batch_matrices)

print(inverted_matrices)

tf.Tensor(
[[[0.86242926 0.54649913 0.33367538 0.6375767 ]
  [0.15247214 0.16092706 0.03879619 0.3351897 ]
  [0.9687394  0.58670473 0.6707103  0.41152215]
  [0.21115458 0.18559194 0.90128255 0.16320539]]

 [[0.3176098  0.5182067  0.53228605 0.905663  ]
  [0.8949578  0.03394866 0.76191616 0.19996047]
  [0.60888755 0.23472798 0.86348295 0.2910093 ]
  [0.7135004  0.7675817  0.5946069  0.54300535]]

 [[0.5452622  0.7498468  0.20748651 0.48811293]
  [0.06483603 0.9891709  0.2525226  0.33118272]
  [0.8436744  0.12197447 0.48864317 0.263489  ]
  [0.29201293 0.98951495 0.6710489  0.9000653 ]]], shape=(3, 4, 4), dtype=float32)


tf.Tensor(
[[[ 1.31953373e+01 -1.54361601e+01 -8.79151535e+00  2.32165313e+00]
  [-2.88347244e+01  3.09110870e+01  2.24021320e+01 -7.32637691e+00]
  [ 1.45680416e+00 -1.91324770e+00 -1.35837376e+00  1.66340995e+00]
  [ 7.67281151e+00 -4.61416054e+00 -6.59909296e+00  2.26883698e+00]]

 [[-1.73119947e-01  2.16992092e+00 -2.30848122e+00  7.26842642e-01]
  [-9.68229115e-01 

### **Exercise 1.47**: Create two tensors of shape `(3, 4)` with random values. Concatenate these tensors along a new axis.

```python
tensor1 = tf.random.uniform((3, 4))
tensor2 = tf.random.uniform((3, 4))

# Concatenate along a new axis
concatenated = tf.stack([tensor1, tensor2], axis=0)

print(concatenated)
```

In [ ]:
tensor1 = tf.random.uniform((3, 4))
print(tensor1)
print("\n")
tensor2 = tf.random.uniform((3, 4))
print(tensor2)
print("\n")
# Concatenate along a new axis
concatenated = tf.stack([tensor1, tensor2], axis=0)
print(concatenated)

tf.Tensor(
[[0.73910487 0.28574753 0.1397177  0.6779921 ]
 [0.45329213 0.49542665 0.5946841  0.96912897]
 [0.44934607 0.5424191  0.56482923 0.5133703 ]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[0.18543243 0.33425295 0.33643305 0.20439816]
 [0.4855467  0.10545135 0.0451051  0.42295265]
 [0.48240793 0.05822515 0.0625062  0.73378444]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[[0.73910487 0.28574753 0.1397177  0.6779921 ]
  [0.45329213 0.49542665 0.5946841  0.96912897]
  [0.44934607 0.5424191  0.56482923 0.5133703 ]]

 [[0.18543243 0.33425295 0.33643305 0.20439816]
  [0.4855467  0.10545135 0.0451051  0.42295265]
  [0.48240793 0.05822515 0.0625062  0.73378444]]], shape=(2, 3, 4), dtype=float32)


### **Exercise 1.48**: Create a tensor of shape `(4, 4)` with random values. Compute the determinant of this matrix.

```python
matrix = tf.random.uniform((4, 4))

# Compute determinant
determinant = tf.linalg.det(matrix)

print(determinant)
```

In [ ]:
matrix = tf.random.uniform((4, 4))

# Compute determinant
determinant = tf.linalg.det(matrix)

print(determinant)

tf.Tensor(-0.32389793, shape=(), dtype=float32)


### **Exercise 1.49**: Create a batch of 3 positive-definite matrices, each of shape `(4, 4)`, with random values. Perform Cholesky decomposition on each matrix in the batch.

```python
# Batch of positive-definite matrices
batch_matrices = tf.random.uniform((3, 4, 4))
batch_matrices = tf.matmul(batch_matrices, batch_matrices, transpose_b=True)

# Cholesky decomposition
cholesky_decomp = tf.linalg.cholesky(batch_matrices)
print(cholesky_decomp)
```


In [ ]:
# Batch of positive-definite matrices
batch_matrices = tf.random.uniform((3, 4, 4))
print(batch_matrices)
print("\n")
batch_matrices = tf.matmul(batch_matrices, batch_matrices, transpose_b=True)
print(batch_matrices)
print("\n")
# Cholesky decomposition
cholesky_decomp = tf.linalg.cholesky(batch_matrices)
print(cholesky_decomp)

tf.Tensor(
[[[0.3212211  0.55865955 0.32018745 0.88594365]
  [0.40889728 0.4788828  0.59713125 0.46238065]
  [0.16048479 0.53426135 0.18632841 0.6329986 ]
  [0.49749982 0.07061589 0.55024374 0.10969508]]

 [[0.0410322  0.17499316 0.20238101 0.6626613 ]
  [0.06052661 0.42403042 0.5485209  0.30200505]
  [0.52050006 0.9704243  0.18596363 0.96611476]
  [0.7981188  0.70230055 0.22502637 0.08154845]]

 [[0.01508486 0.13607883 0.6716677  0.37769198]
  [0.19850993 0.36502314 0.68567073 0.6758326 ]
  [0.7761085  0.47125196 0.98519874 0.6962539 ]
  [0.6266335  0.48116314 0.45561588 0.36368763]]], shape=(3, 4, 4), dtype=float32)


tf.Tensor(
[[[1.3026996  0.99971604 0.9704824  0.47262245]
  [0.99971604 0.96688735 0.72541916 0.61653167]
  [0.9704824  0.72541916 0.7465961  0.28953135]
  [0.47262245 0.61653167 0.28953135 0.5672938 ]]

 [[0.51238436 0.38782325 0.86901724 0.25522643]
  [0.38782325 0.5755475  0.83677    0.49416393]
  [0.86901724 0.83677    2.180604   1.2175823 ]
  [0.25522643 0.4941639

### **Exercise 1.50**: Create a tensor of shape `(3, 4)` with random values. Compute the cumulative sum of this tensor along the second axis.

```python
tensor = tf.random.uniform((3, 4))

# Cumulative sum along axis 1
cumsum = tf.math.cumsum(tensor, axis=1)

print(cumsum)
```

In [ ]:
tensor = tf.random.uniform((3, 4))
print(tensor)
print("\n")
# Cumulative sum along axis 1
cumsum = tf.math.cumsum(tensor, axis=1)

print(cumsum)

tf.Tensor(
[[0.39759624 0.36899745 0.3430133  0.56731474]
 [0.05217099 0.10449576 0.96972203 0.3056463 ]
 [0.9197496  0.7126144  0.8989712  0.3569312 ]], shape=(3, 4), dtype=float32)


tf.Tensor(
[[0.39759624 0.7665937  1.109607   1.6769217 ]
 [0.05217099 0.15666676 1.1263888  1.4320351 ]
 [0.9197496  1.632364   2.5313354  2.8882666 ]], shape=(3, 4), dtype=float32)
